File: backend/src/api/dto/leaderboard.py
"""Leaderboard DTOs (see ScoresService.get_leaderboard) plus the daily-challenge
leaderboard (see ScoresService.get_daily_leaderboard)."""

from datetime import date
from typing import Literal
from uuid import UUID

from pydantic import BaseModel

from services.scores_service import DailyLeaderboardEntry, LeaderboardEntry

LeaderboardWindow = Literal["all", "weekly", "daily"]


class LeaderboardEntryOut(BaseModel):
    rank: int
    username: str
    skin_person_id: UUID | None
    best_score: int

    @classmethod
    def from_entry(cls, entry: LeaderboardEntry) -> "LeaderboardEntryOut":
        return cls(
            rank=entry.rank,
            username=entry.username,
            skin_person_id=entry.skin_person_id,
            best_score=entry.best_score,
        )


class LeaderboardOut(BaseModel):
    window: LeaderboardWindow
    entries: list[LeaderboardEntryOut]

    @classmethod
    def from_entries(cls, window: LeaderboardWindow, entries: list[LeaderboardEntry]) -> "LeaderboardOut":
        return cls(window=window, entries=[LeaderboardEntryOut.from_entry(e) for e in entries])


class DailyLeaderboardEntryOut(LeaderboardEntryOut):
    # None on any date other than today - see ScoresService.DailyLeaderboardEntry.
    streak: int | None

    @classmethod
    def from_entry(cls, entry: DailyLeaderboardEntry) -> "DailyLeaderboardEntryOut":
        return cls(
            rank=entry.rank,
            username=entry.username,
            skin_person_id=entry.skin_person_id,
            best_score=entry.best_score,
            streak=entry.streak,
        )


class DailyLeaderboardOut(BaseModel):
    date: date
    entries: list[DailyLeaderboardEntryOut]

    @classmethod
    def from_entries(cls, challenge_date: date, entries: list[DailyLeaderboardEntry]) -> "DailyLeaderboardOut":
        return cls(date=challenge_date, entries=[DailyLeaderboardEntryOut.from_entry(e) for e in entries])

File: frontend/src/api/types/leaderboard.ts
// Leaderboards (requires login, unlike the personal records in records.ts) -
// mirrors backend/src/api/dto/leaderboard.py.

export type LeaderboardWindow = "all" | "weekly" | "daily"

export interface LeaderboardEntryOut {
  rank: number
  username: string
  skin_person_id: string | null
  best_score: number
}

export interface LeaderboardOut {
  window: LeaderboardWindow
  entries: LeaderboardEntryOut[]
}

// Same entry shape as the normal LeaderboardOut plus a streak, scoped to one specific day's
// challenge instead of a rolling window. The streak is null on any date other than today - the
// backend only computes it for the current day, since that's the only one showing the badge.
export interface DailyLeaderboardEntryOut extends LeaderboardEntryOut {
  streak: number | null
}

export interface DailyLeaderboardOut {
  date: string
  entries: DailyLeaderboardEntryOut[]
}
